### DISCUSSION TITLE
A CT dynamic growth model with ctsem

### DISCUSSION THREAD

**Discussion by JulianFLohmann** [cdriveraus/ctsem#17] [Q&A]
Dear Charlie, dear community,
for my masters thesis, we are trying to set up some kind of ct dynamic growth model, so a combination of standard linear growth model and ct dynamic model. We tried some specifications and got the best model parameters using the following model:

CT_GM <- ctModel( type = 'stanct' ,
                    n.manifest = 1 ,
                    n.latent = 2 ,
                    manifestNames = c( ‘Y1’ ) , latentNames = c(  'f_Y1' , 'lin' ) ,
                    DRIFT = matrix( c( 'a', 0 , 0 , 0) , nrow = 2,ncol=2 ) ,
                    DIFFUSION = matrix( c( 'q' , 0 , 0 , 0 ) , nrow = 2 , ncol = 2 ) ,
                    T0VAR = matrix( c( 0 , 0 , 0,  't0var'), nrow = 2 , ncol = 2 ), 
                    CINT = matrix( c( 0 , 'b_lin' ) , nrow = 2 , ncol = 1 ) ,
                    T0MEANS = matrix( c( 0 , 't0m' ) , nrow = 2 , ncol = 1 ) ,
                    LAMBDA = matrix( c( 1 , 1 ) , nrow = 1 , ncol = 2 ) ,
                    MANIFESTMEANS = matrix( 0 , ncol = 1 ) ,
                    MANIFESTVAR = matrix( c( 0 ) , nrow = 1 , ncol =1 )   )

So we have the standard growth curve  parameters in the second model and can account for the dynamic structure of the residuals in the first. We are also interested in the person parameters and found the following problem:  In this model, the person specific intercepts are set to the observed scores – this results in wrong individual trajectories. So we assume our model is misspecified. 
Question: Is it possible to specify a time-point specific error variance for T0 in the ctsem framework - some kind of MANIFESTVAR just for T0 (as maybe one possible solution for the problem)? 
Or more general, do you know if there is a possibility to set up a CALT Model (Delsing & Oud, 2008) in ctsem? Do you have any suggestions what we could try?

Best regards,
Julian

**Comment by cdriveraus** [comment]
Hi Julian, what do you mean by 'the person specific intercepts are set to the observed scores' ? Do you allow individual specific parameters in this model? You could also post up the pdf from ctModelLatex(CT_GM) 

What you ask re a time point specific measurement error *is* possible but it's a little tricky and relies on some undocumented tricks -- basically you would include a time dependent predictor which is 1 at the first observation and 0 after that, and set MANIFESTVAR = 'log1p_exp(baseSD + T0SDeffect * tdpreds[rowi,1])', PARS= c('baseSD','T0SDeffect') . However, this is slower because it will have to compile, and I actually doubt this is what you really want -- why do you think you need this? 

I don't remember the CALT model specifically but I'm relatively certain you can specify it. The way you have things structured allows a straightforward splitting between the general growth trend and the dynamics, which can definitely be nice in some cases, but in other cases it's actually better to combine the two, and model the second process as essentially a time varying intercept parameter, which has random noise (i.e. a diffusion parameter), impacts the first process with a fixed value of 1, and the first process has an estimated cross effect parameter to the second, because usually for higher levels of the observed process, the rate of growth decreases (or vice versa for decline). You can see an example of such a specification in the sunspots example in ctDocs() or the quick start at https://cdriver.netlify.app/post/ctsem-quick-start/

Perhaps this is enough, or maybe you want to describe your problem and what you want a little more broadly?

**Reply by JulianFLohmann** [reply]
Dear Charlie, thank you very much for your answer!

Yes, I’m sorry, I didn’t mention that – I set T0M and CINT to ‘individual varying’ 
[ModelLatex_CTGM.pdf](https://github.com/cdriveraus/ctsem/files/6707179/ModelLatex_CTGM.pdf)
.
What I’m thinking about this problem at the moment is that because I set T0M and T0Var in the first process to 0 and also the MANIFESTVAR to 0, T0-Residuals for the second (the linear process) can not occur. That leads to the fact that the subject specific T0M Intercepts are equal to the observed scores of T0 in that model (what results in misleading individual trajectories, as already mentioned.)
Thanks for the trick, but I was not able to run this until now. But I’m also not sure if that’s really a proper solution for my problem.
I was thinking about changing the line T0VAR to:
T0VAR = matrix( c( 't0residuals' , 0 , 0 , 't0var'), nrow=2,ncol=2 ) ,
But maintaining the lines:
T0MEANS=matrix( c( 0 , 't0m' ) , nrow=2,ncol=1 ) ,
MANIFESTVAR=matrix( c( 0 ) , nrow=1,ncol=1 )
[ModelLatex_CTGM_2_with T0residual.pdf](https://github.com/cdriveraus/ctsem/files/6707185/ModelLatex_CTGM_2_with.T0residual.pdf)

Could that be a right way to an appropriate solution for my desired model? The results look quite better, and I will check the model more closely now...

In the CALT model (Delsing & Oud, 2008), the first measurement occasion (T0) is treated as exogenous and predicts T1. T0 does not load on the intercept and slope factor but is allowed to covary with them. In principle, do you think it’s possible to make that model work in ctsem?

Best
Julian

**Reply by cdriveraus** [reply]
Ok. In such models it does get a little tricky to think about the interaction between your T0VAR specification and the individual variation. What is really happening in the background when you set individually varying parameters, is that the latent system is expanded according to the number of varying parameters specified, and the dynamics / measurement properties of the expanded system are fixed to zero, and the variance / covariance of these expanded elements are freely estimated. The only difference is for T0MEANS, when they are set indvarying the system is not expanded, but overwrites any T0VAR specification (on the particular elements / covariances) that you have put in, because it is required for identification, is computationally more efficient, and then helps identify other individual variation because of the covariance with first time point. That possibly sounds rather complicated, but basically the key point is that your T0VAR specification is  anyway not being used. If you want to be clearer about the exact structure you have, in this case (because you only have individual variation in intercept type parameters) you can turn off individual variation and just rely on your own setup. 

All that said, setting T0VAR for the first process free like you propose does seem like the most appropriate solution. However, why do you fix measurement error to zero? This imposes pretty strong assumptions on the model --if measurement error is zero, and you only have one element of variability (the t0means of the growth process) at that point, then yes, those t0means will become equivalent to your first observation. If it was my model, I'd free the t0var like proposed *and* the measurement error term. Otherwise your dynamic process will be some kind of hybrid between the random, non-temporally dependent uncertainty of measurement, and the growing-in-time uncertainty of dynamics.

**Reply by JulianFLohmann** [reply]
I assumed that I have to make the assumption of no measurement error, because I do not specify a measurement model haven’t I?
And another question: when I use optimizer=TRUE, nopriors=TRUE in ctstan I get ML estimation – is it still a Bayesian model?

**Reply by cdriveraus** [reply]
You always have some kind of measurement model -- in your case you are saying that the observations you have are a perfect reflection of a first order linear dynamic system + growth curve. Even if you *have* measured your system perfectly, it's unlikely that the system you specify is the *true system*, and allowing measurement error means that information from further back than just the single preceding observation can also be used to predict the next point. Freeing the measurement error term is identified by restricting it in some way, in the default case ctsem just assumes measurement error is constant over time, though I've an example on the blog where this too can be modified.

You could still view it as a Bayesian model with flat priors estimated using maximum a posteriori estimation, but most people would just call it maximum likelihood.

**Reply by JulianFLohmann** [reply]
Charles, thank you so much for the clarifications and helpful suggestions!! Great package, it’s a lot of fun for me to work with ctsem!!
